File: 2024-04-17/Binary_Tree_Preorder_Traversal.java
//  144. Binary Tree Preorder Traversal
//
//  https://leetcode.com/problems/binary-tree-preorder-traversal/

import java.util.*;

public class Binary_Tree_Preorder_Traversal {
    public static List<Integer> ans;
    public static void solve(TreeNode root) {
        if(root == null) return;
        ans.add(root.val);
        solve(root.left);
        solve(root.right);
    }
    public static List<Integer> preorderTraversal(TreeNode root) {
        ans = new ArrayList<>();
        solve(root);
        return ans;
    }
    public static void main(String[] args) {
        Integer[] arr = new Integer[]{1,2,3,4,5,6,7};
        Tree tree = new Tree();
        TreeNode root = tree.arrayToTree(arr);
        List<Integer> ans = preorderTraversal(root);
        System.out.println(ans.toString());
    }
}

File: 2026-02-12/Binary_Tree_Preorder_Traversal.py
from typing import Optional, List
from Tree import TreeNode, Tree


class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        result = []
        cur = root

        while cur:
            if not cur.left:
                result.append(cur.val)
                cur = cur.right
            else:
                leftPointer = cur.left
                while leftPointer.right and leftPointer.right != cur:
                    leftPointer = leftPointer.right

                if not leftPointer.right:
                    result.append(cur.val)
                    leftPointer.right = cur
                    cur = cur.left
                else:
                    leftPointer.right = None
                    cur = cur.right

        return result


sol = Solution()
tree = Tree()
node = tree.array_to_tree([1,2,3,4,5,None,8,None,None,6,7,9])
print(sol.preorderTraversal(node))
